src/server/coprocessor/executor/base_scanner.rs

```diff
@@ -0,0 +1,394 @@
+                2 => Datum::Bytes(b"abc".to_vec()),
+                3 => Datum::Dec(10.into())
+            ];
+            let mut encode_value = HashMap::default();
+            let col_ids: Vec<_> = row.iter().map(|(&id, _)| id).collect();
+            let col_values: Vec<_> = row.iter()
+                .map(|(k, v)| {
+                    let f = flatten(v.clone()).unwrap();
+                    let value = datum::encode_value(&[f]).unwrap();
+                    encode_value.insert(*k, value);
+                    v.clone()
+                })
+                .collect();
+
+            let value = table::encode_row(col_values, &col_ids).unwrap();
+            let mut buf = vec![];
+            buf.encode_i64(handle as i64).unwrap();
+            let key = table::encode_row_key(table_id, &buf);
+            if pk.is_empty() {
+                pk = key.clone();
+                pk_handle = handle as i64;
+            }
+            encode_data.push(encode_value);
+            data.push((key, value));
+        }
+        Data {
+            data: data,
+            pk: pk,
+            pk_handle: pk_handle,
+            encode_data: encode_data,
+            cols: cols,
+        }
+    }
+
+    const START_TS: u64 = 10;
+    const COMMIT_TS: u64 = 20;
+
+    pub struct TestStore {
+        pk: Vec<u8>,
+        snapshot: Box<Snapshot>,
+        ctx: Context,
+        engine: Box<Engine>,
+    }
+
+    impl TestStore {
+        pub fn new(data: &[(Vec<u8>, Vec<u8>)], pk: Vec<u8>) -> TestStore {
+            let engine = engine::new_local_engine(TEMP_DIR, ALL_CFS).unwrap();
+            let ctx = Context::new();
+            let snapshot = engine.snapshot(&ctx).unwrap();
+            let mut store = TestStore {
+                pk: pk,
+                snapshot: snapshot,
+                ctx: ctx,
+                engine: engine,
+            };
+            store.init_data(data);
+            store
+        }
+
+        fn init_data(&mut self, data: &[(Vec<u8>, Vec<u8>)]) {
+            let mut statistics = Statistics::default();
+            // do prewrite.
+            {
+                let mut txn = MvccTxn::new(self.snapshot.as_ref(), &mut statistics, START_TS, None);
+                for &(ref key, ref value) in data {
+                    txn.prewrite(Mutation::Put((make_key(key), value.to_vec())),
+                                  &self.pk,
+                                  &Options::default())
+                        .unwrap();
+                }
+                self.engine.write(&self.ctx, txn.modifies()).unwrap();
+            }
+            self.refresh_snapshot();
+            // do commit
+            {
+                let mut txn = MvccTxn::new(self.snapshot.as_ref(), &mut statistics, START_TS, None);
+                for &(ref key, _) in data {
+                    txn.commit(&make_key(key), COMMIT_TS).unwrap();
+                }
+                self.engine.write(&self.ctx, txn.modifies()).unwrap();
```

Comment:
IMO It seems better to wrap `engine.write()` and `refresh_snapshot()` into a helper function like
```
fn write_txn(txn ...) {
    engine.write(txn),
    self.snapshot = self.engine.snapshot(...
}
```
and then call `write_txn` here.